src/main/java/com/limechain/network/protocol/transactions/TransactionsEngine.java

```diff
@@ -0,0 +1,142 @@
+package com.limechain.network.protocol.transactions;
+
+import com.limechain.network.ConnectionManager;
+import com.limechain.network.protocol.transactions.scale.TransactionsReader;
+import com.limechain.network.protocol.transactions.scale.TransactionsWriter;
+import com.limechain.sync.warpsync.SyncedState;
+import io.emeraldpay.polkaj.scale.ScaleCodecReader;
+import io.emeraldpay.polkaj.scale.ScaleCodecWriter;
+import io.libp2p.core.PeerId;
+import io.libp2p.core.Stream;
+import lombok.extern.java.Log;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.logging.Level;
+
+/**
+ * Engine for handling transactions on Transactions streams.
+ */
+@Log
+public class TransactionsEngine {
+    private static final int HANDSHAKE_LENGTH = 1;
+
+    private final ConnectionManager connectionManager = ConnectionManager.getInstance();
+    private final SyncedState syncedState = SyncedState.getInstance();
+
+    /**
+     * Handles an incoming request as follows:
+     *
+     * <p><b>On streams we initiated:</b>  adds streams, where we receive a handshake message,
+     * to initiator streams in peer's {@link com.limechain.network.dto.PeerInfo} , ignores all other message types.
+     *
+     * <p><b>On responder stream: </b>
+     * <p>If message payload contains a valid handshake, adds the stream when the peer is not connected already,
+     * ignore otherwise. </p>
+     * <p>On transactions messages {@link SyncedState}. </p>
+     * <p>Logs and ignores other message types. </p>
+     *
+     * @param message received message as byre array
+     * @param peerId  peer id of sender
+     * @param stream  stream, where the request was received
+     */
+    public void receiveRequest(byte[] message, PeerId peerId, Stream stream) {
+        System.out.println("Transaction message length:" + message.length);
+        if (message == null) {
```

Comment:
Null check should be done before using `message.length`.
`message[0]` will also throw a NPE in the log.